File: dsa_leetcode_beginner/0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) {
        this.val = val;
    }
}

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode slow = dummy;
        ListNode fast = head;

        // Move fast pointer so that the gap between slow and fast is n nodes apart
        for (int i = 0; i < n; i++) {
            fast = fast.next;
        }

        // Move fast to the end, maintaining the gap
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }

        // Skip the desired node
        slow.next = slow.next.next;

        return dummy.next;
    }
}

File: dsa_leetcode_beginner/0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0) # Create a dummy node to simplify edge cases
        dummy.next = head
        slow, fast = dummy, head
        
        # Move fast pointer so that the gap between slow and fast is n nodes apart
        for _ in range(n):
            fast = fast.next
        
        # Move fast to the end, maintaining the gap
        while fast:
            fast = fast.next
            slow = slow.next
        
        # Skip the desired node
        slow.next = slow.next.next

        # Return head of the modified list
        return dummy.next  
    

#Question: https://leetcode.com/problems/remove-nth-node-from-end-of-list
#Blog: https://blog.unwiredlearning.com/remove-nth-node-from-end-of-list